internal/api/impl/auth/builtin.go

```diff
@@ -0,0 +1,81 @@
+package auth
+
+import (
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+	"github.com/perses/perses/internal/api/interface/v1/user"
+	"github.com/perses/perses/internal/api/shared"
+	"github.com/perses/perses/internal/api/shared/crypto"
+	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	"github.com/perses/perses/pkg/model/api"
+	"github.com/sirupsen/logrus"
+)
+
+type builtinEndpoint struct {
+	dao user.DAO
+	jwt crypto.JWT
+}
+
+func NewBuiltinEndpoint(dao user.DAO, jwt crypto.JWT) Endpoint {
+	return &builtinEndpoint{
+		dao: dao,
+		jwt: jwt,
+	}
+}
+
+func (e *builtinEndpoint) CollectRoutes(g *shared.Group) {
+	g.POST("/auth/login", e.auth, true)
+}
+
+func (e *builtinEndpoint) auth(ctx echo.Context) error {
+	body := &api.Auth{}
+	if err := ctx.Bind(body); err != nil {
+		return shared.HandleBadRequestError(err.Error())
+	}
+	usr, err := e.dao.Get(body.Login)
+	if err != nil {
+		if databaseModel.IsKeyNotFound(err) {
+			return shared.HandleBadRequestError("wrong login or password ")
+		}
+	}
+
+	if !crypto.ComparePasswords(usr.Spec.Password, body.Password) {
+		return shared.HandleBadRequestError("wrong login or password ")
+	}
+	accessToken, err := e.accessToken(ctx, body.Login)
+	if err != nil {
+		return err
+	}
+	refreshToken, err := e.refreshToken(ctx, body.Login)
+	if err != nil {
+		return err
+	}
+
+	return ctx.JSON(http.StatusOK, api.AuthResponse{
+		AccessToken:  accessToken,
+		RefreshToken: refreshToken,
+	})
+}
+
+func (e *builtinEndpoint) accessToken(ctx echo.Context, login string) (string, error) {
```

Comment:
since `accessToken` and `refreshToken` are common in all implementation, it would worth to create a common struct that hold this implementation.

Like


```golang
type tokenManagement struct {
  jwt crypto.JWT
}

func (e *tokenManagement) accessToken(ctx echo.Context, login string) (string, error) {
	accessToken, err := e.jwt.SignedAccessToken(login)
	if err != nil {
		logrus.WithError(err).Errorf("unable to generate the access token")
		return "", shared.InternalError
	}
	jwtHeaderPayloadCookie, signatureCookie := e.jwt.CreateAccessTokenCookie(accessToken)
	ctx.SetCookie(jwtHeaderPayloadCookie)
	ctx.SetCookie(signatureCookie)
	return accessToken, nil
}

func (e *tokenManagement) refreshToken(ctx echo.Context, login string) (string, error) {
	refreshToken, err := e.jwt.SignedRefreshToken(login)
	if err != nil {
		logrus.WithError(err).Errorf("unable to generate the refresh token")
		return "", shared.InternalError
	}
	ctx.SetCookie(e.jwt.CreateRefreshTokenCookie(refreshToken))
	return refreshToken, nil
}
```

and then in the different endpoint

```golang
type builtinEndpoint struct {
	dao user.DAO
	jwt crypto.JWT
        tokenManagement *tokenManagement
}

func (e *builtinEndpoint) auth(ctx echo.Context) error {
	body := &api.Auth{}
	if err := ctx.Bind(body); err != nil {
		return shared.HandleBadRequestError(err.Error())
	}
	usr, err := e.dao.Get(body.Login)
	if err != nil {
		if databaseModel.IsKeyNotFound(err) {
			return shared.HandleBadRequestError("wrong login or password ")
		}
	}

	if !crypto.ComparePasswords(usr.Spec.Password, body.Password) {
		return shared.HandleBadRequestError("wrong login or password ")
	}
	accessToken, err := e.tokenManagement.accessToken(ctx, body.Login)
	if err != nil {
		return err
	}
	refreshToken, err := e.tokenManagement.refreshToken(ctx, body.Login)
	if err != nil {
		return err
	}

	return ctx.JSON(http.StatusOK, api.AuthResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
	})
}
```

Reply:
Sure the thing is the accessToken/refreshToken is managed by the OIDC library with good old http writer. No problem the duplication has already been identified, I will work on that later

Reply:
Splitted the token management in two steps : the creation of token/cookies (framework independant) and then the set of the cookies (framework dependent so duplicated for now)